### ISSUE TITLE
Problem reproducing paper results

### ISSUE THREAD

**Issue by justanothercoder** [vfleaking/rnnprop#1]
Hello! Thank you for the interesting paper!

I am trying to reproduce some results from the paper. As far as I understood all the parameters, which are set by default, are the same as in the paper. However, the number of training epochs is set to 0 i.e. train until keyboard interruption. I've tried to train for about 300 epochs on `mnist-nn-sigmoid-100` (at this moment printed average loss of rnnprop became lower than the loss of GD). However, test performance on the same problem wasn't quite the same as in the paper. Sometimes rnnprop even diverged.

Could you please share your thoughts on how to reproduce results from the paper exactly?
Thanks.

**Comment by shunhuajiang** [comment]
Hi!

Thank you for reading our paper and implementing our codes. Here are some reasons I think that might lead to different results in your reproduction.
1. The most likely reason is that all the results shown in our paper is averaged after multiple runs. In details, all the numbers and figures are the average results of 100 runs. We notice that there exists randomness when training neural networks (since each time we use different mini-batches of data). This randomness also exists for traditional optimization algorithms, e.g., Adam. If in your experiments, you only run the tests for a few times, then it is possible to have slightly different results from those in the paper.
2. The difference might also come from our different ways to choose which epoch of the model to use. As we mentioned in the paper, we chose the final epoch number where the moving average of the train loss (loss on `mnist-nn-sigmoid-100`) is lowest. In details, we use the moving average of the losses of the past 1000 iterations (note that there are 100 iterations in 1 epoch) with decay factor 0.9. The moving average is to estimate the expected train loss of the model at the end of 10 epoch. The numbers 0.9, 1000 are chosen somewhat arbitrarly, so changing them will not lead to much difference.
3. Two tips that might be useful: You can set the `frequency` in the `test_list.py` to be 1 so that this test task is tested in every epoch. In this way maybe you can find out what went wrong in the process of training the optimizer. Also, you can change the label `n_tests` at the end of `main.py` to be larger than 1, e.g., 10, in this way in the end of each epoch the test tasks are run for multiple times and compute the average.

Also, I note that you said 
> Sometimes rnnprop even diverged.

Could you please provide more details? As in our experiments we trained 10 optimizers and we did not observe any divergence when running on the test tasks mentioned in our paper. We would appreciate more detailed descriptions of your experiments.

Thank you!

**Comment by justanothercoder** [comment]
Thanks, I will try what you suggest about the experiments and write later what I've got.

Concerning rnnprop divergence: after training for 600 epochs on `mnist-nn-sigmoid-100` I've observed divergence on `mnist-nn-l2-elu-100`. [Here](http://i.imgur.com/0UHLbnQ.png) are the test results which are averaged after 3 runs (that is a part of the plot with moving average of function values). Blue curve corresponds to the model I am trying to train, others are Adam and RNNProp is the last one. The number of iterations is 1000. Optimizee parameters are defaults from your code.

**Comment by shunhuajiang** [comment]
Hi!

We’ve done a little bit of experiments that might be helpful.

We tested 10 trained RNNprops (including the one which we used to run all the experiments in the paper) on ` mnist-nn-l2-elu-100 `. [Here](http://imagizer.imageshack.us/a/img924/8846/h6TRic.png) is the result. As you can see, overall they perform all right, though there is certain randomness in the behaviors of the trained models and it is possible for you to get slightly different results from the paper.

The number of training epochs of these 10 models are selected the same way as I mentioned before. We also observe that the selected epoch of these 10 models is around 200~300, and the training loss (the loss on ` mnist-nn-sigmoid-100 `) is around 0.47 at the selected epoch. By choosing a similar epoch number for your model, you should be able to get a similar model to the ones we trained.

**Comment by justanothercoder** [comment]
Thank you very much! 

Actually, I've found some quite subtle mistakes in _my_ code which were preventing RNNProp from working properly. Everything is good at the moment.  Now I have a good baseline :)
